argo-cd-apps/base/enterprise-contract.yaml

```diff
@@ -1,77 +1,52 @@
-apiVersion: argoproj.io/v1alpha1
-kind: Application
-metadata:
-  name: rekor
-
-spec:
-  project: default
-  source:
-    path: helm-charts/rekor
-    repoURL: 'https://github.com/sigstore/sigstore-helm-operator'
-    targetRevision: "143b8d71fc861d7c7ac93a15d0fbd33654b75908"
-
-    helm:
-      valueFiles:
-        - values.yaml
-      values: |-
-        server:
-          strategy:
-            type: Recreate
-          ingress:
-            hostname: rekor-server.enterprise-contract-service.svc
-            annotations:
-              route.openshift.io/termination: "edge"
-        mysql:
-          auth:
-            password: "password"
-
-  destination:
-    namespace: enterprise-contract-service
-    server: https://kubernetes.default.svc
-
-  syncPolicy:
-    # Comment this out if you want to manually trigger deployments (using the
-    # Argo CD Web UI or Argo CD CLI), rather than automatically deploying on
-    # every new Git commit to your directory.
-    automated:
-      prune: true
-      selfHeal: true
-
-    syncOptions:
-      - CreateNamespace=true
-
-    retry:
-      limit: 50 # number of failed sync attempt retries; unlimited number of attempts if less than 0
-      backoff:
-        duration: 15s # the amount to back off. Default unit is seconds, but could also be a duration (e.g. "2m", "1h")
 ---
 apiVersion: argoproj.io/v1alpha1
-kind: Application
+kind: ApplicationSet
 metadata:
   name: enterprise-contract
-
+  namespace: openshift-gitops
 spec:
-  project: default
-
-  source:
-    path: components/enterprise-contract
-    repoURL: https://github.com/redhat-appstudio/infra-deployments.git
-    targetRevision: main
-
-  destination:
-    namespace: enterprise-contract-service
-    server: https://kubernetes.default.svc
-
+  generators:
+    - list:
+        elements:
+          # Determines which Rekor (Tekton Chains transparency log) variant will
+          # be used, can be:
+          #  * off      - transparency off
+          #  * local    - transparency on, with an in-cluster Rekor instance
+          #  * sigstore - transparency on, with Sigstore public Rekor instance
+          - rekor: off
```

Comment:
What happens if I do this?
```yaml
  generators:
    - list:
        elements:
          - rekor: off
          - rekor: local
          - rekor: sigstore
```

Reply:
Don’t, ArgoCD would infinity loop over the chains-config ConfigMap

Reply:
I think that's a pretty good indicator that this is not the right tool for the job...

All we need is a single Application to deploy Rekor locally. The other modes are not deploying Rekor. They're just manipulating the Chains configuration. I also think it should be perfectly valid to have Rekor running, but have Chains not using it, especially during development and debugging phases. I really disagree with coupling the Rekor deployment to the Chains configuration management. The Rekor service shouldn't care if anyone is using it, that's a Chains concern.

We should have a specific Chains config and specific mode for Rekor in staging. For development/preview mode we could have a default, and leverage an env var in preview.env to switch the Rekor mode.

I still don't see the benefits of using an ApplicationSet. In fact, it seems like it makes it easier for a developer to shoot themselves in the foot.

Reply:
I don't think there is any value in having Rekor running without Tekton Chains using it. After all if this is something that one wishes to have, what's stopping them from deploying Rekor manually? _Validity_ of this is more of a personal preference, for Application Studio there is no other use for Rekor other than to store entries created by Tekton Chains. And that use case is, as far as I understand, still under debate. If there were any other use cases for Rekor, they would surely be voiced by now.

The benefit of using ApplicationSet is that it allows switching between these three different modes without having anything leftover or inconsistent upon making the switch. There is no other way that I know of doing this with ArgoCD.

If the debate is around having three modes, usefulness of them or technical implementation I would welcome this, but this to me seems rather pointless to debate.

Reply:
My point is that Rekor is a standalone component. At its core, Rekor couldn't care less if Chains is using it or not. On the other hand, Chains should be aware of Rekor if it intends to use it. This is the natural relationship between these two components. I think we should avoid murking that relationship.

My other point is that by using ApplicationSets in this case, it's very easy for someone to do something that puts their cluster in an unstable position, as you mentioned earlier. I think we should avoid knowingly adding such things.

Also, I disagree that toggling off Rekor should get rid of the deployment because this is yet another way to mess things up. What if someone decides that we should turn off Rekor for a little bit, for whatever reason? It would be a bad experience if that also deleted data. I think once Rekor is deployed, any sort of removal should be manual.

Reply:
> My point is that Rekor is a standalone component. At its core, Rekor couldn't care less if Chains is using it or not. On the other hand, Chains should be aware of Rekor if it intends to use it. This is the natural relationship between these two components. I think we should avoid murking that relationship.

In theory it is, in practice it is not. Most definitely by everything we are doing here Rekor is tied to Chains. There is no Rekor without Chains.

> My other point is that by using ApplicationSets in this case, it's very easy for someone to do something that puts their cluster in an unstable position, as you mentioned earlier. I think we should avoid knowingly adding such things.

I don't follow this argument, it is very easy to make other mistakes that put cluster in unstable state. We should not be bound by what can be done by mistake but what we enable as functionality.

> Also, I disagree that toggling off Rekor should get rid of the deployment because this is yet another way to mess things up. What if someone decides that we should turn off Rekor for a little bit, for whatever reason? It would be a bad experience if that also deleted data. I think once Rekor is deployed, any sort of removal should be manual.

This argument doesn't make sense to me. It is not intended for this to be used to "turn off Rekor for a little bit". The intent for this is to deploy and configure the use of Rekor when the cluster is provisioned. By this line of thinking what if someone decides to delete all storage of a cluster _for a little bit for whatever reason_.